lib/onelogin/ruby-saml/response.rb

```diff
@@ -200,27 +487,58 @@ def xpath_first_from_signed_assertion(subelt=nil)
         node
       end
 
+      # Extracts all the appearances that matchs the subelt (pattern)
+      # Search on any Assertion that is signed, or has a Response parent signed
+      # @param subelt [String] The XPath pattern
+      # @return [Array of REXML::Element] Return all matches
+      #
+      def xpath_from_signed_assertion(subelt=nil)
+        node = REXML::XPath.match(
+            document,
+            "/p:Response/a:Assertion[@ID=$id]#{subelt}",
+            { "p" => PROTOCOL, "a" => ASSERTION },
+            { 'id' => document.signed_element_id }
+        )
+        node.concat( REXML::XPath.match(
+            document,
+            "/p:Response[@ID=$id]/a:Assertion#{subelt}",
+            { "p" => PROTOCOL, "a" => ASSERTION },
+            { 'id' => document.signed_element_id }
+        ))
+      end
+
+      # Calculates the fingerprint of the IdP x509 certificate.
+      # @return [String] The fingerprint
+      #
       def get_fingerprint
-        if settings.idp_cert
-          cert = OpenSSL::X509::Certificate.new(settings.idp_cert)
-          Digest::SHA1.hexdigest(cert.to_der).upcase.scan(/../).join(":")
-        else
-          settings.idp_cert_fingerprint
+        settings.idp_cert_fingerprint || begin
+          if settings.idp_cert
+            certificate = OneLogin::RubySaml::Utils.format_cert(settings.idp_cert)
+            x509 = OpenSSL::X509::Certificate.new(certificate)
+            Digest::SHA1.hexdigest(x509.to_der).upcase.scan(/../).join(":")
+          end
         end
       end
 
+      # Validates the Conditions. (If the response was initialized with the :skip_conditions option, this validation is skipped,
+      # If the response was initialized with the :allowed_clock_drift option, the timing validations are relaxed by the allowed_clock_drift value)
+      # If fails, the error is added to the errors array
+      # @param soft [Boolean] soft Enable or Disable the soft mode (In order to raise exceptions when the response is invalid or not)
+      # @return [Boolean] True if satisfies the conditions, otherwise False if soft=True
+      # @raise [ValidationError] if soft == false and validation fails
+      #
       def validate_conditions(soft = true)
         return true if conditions.nil?
         return true if options[:skip_conditions]
 
         now = Time.now.utc
 
-        if not_before && (now + (options[:allowed_clock_drift] || 0)) < not_before
+        if not_before && not_before > (now + (options[:allowed_clock_drift] || 0))
```

Comment:
It would be nice to see this logic encapsulated in a method like `allowed_clock_drift`, at least the `(options[:allowed_clock_drift] || 0)` part.